skiko/src/jvmMain/kotlin/org/jetbrains/skiko/FPSCounter.kt

```diff
@@ -1,43 +1,62 @@
-    private val probability: Double
+    private val periodSeconds: Double,
+    private val showLongFrames: Boolean,
+    private val getLongFrameMillis: () -> Double
 ) {
-    private var i = 0
-    private val times = LinkedList<Double>()
-    private var t1 = System.nanoTime()
-
-    /**
-     * [value] 0.0 - min, 1.0 - max, 0.5 - median
-     */
-    private fun MutableList<Double>.quantile(value: Double) : Double {
-        val index = (value * (size - 1)).toInt()
-        return sorted()[index]
-    }
+    private val times = mutableListOf<Long>()
+    private var lastLogTime = System.nanoTime()
+    private var lastTime = System.nanoTime()
 
     fun tick() {
-        val t2 = System.nanoTime()
-        val frameTime = (t2 - t1) / 1E6
-        t1 = t2
+        val time = System.nanoTime()
+        val timestamp = time.nanosToMillis().toLong()
+        val frameTime = time - lastTime
+        lastTime = time
 
-        i++
         times.add(frameTime)
 
-        if (times.size > count) {
-            times.removeFirst()
+        if (showLongFrames && frameTime > getLongFrameMillis().millisToNanos()) {
+            println("[%d] Long frame %.2f ms".format(timestamp, frameTime.nanosToMillis()))
+        }
+
+        if ((time - lastLogTime) > periodSeconds.secondsToNanos()) {
+            val average = (nanosPerSecond / times.average()).roundToInt()
+            val min = (nanosPerSecond / times.max()!!).roundToInt()
+            val max = (nanosPerSecond / times.min()!!).roundToInt()
+            println("[$timestamp] FPS $average ($min-$max)")
+            times.clear()
+            lastLogTime = time
         }
+    }
+
+    private val nanosPerMillis = 1_000_000.0
+    private val nanosPerSecond = 1_000_000_000.0
+    private fun Long.nanosToMillis(): Double = this / nanosPerMillis
+    private fun Double.millisToNanos(): Long = (this * nanosPerMillis).toLong()
+    private fun Double.secondsToNanos(): Long = (this * nanosPerSecond).toLong()
+}
+
+internal fun defaultFPSCounter(
+    component: Component
+): FPSCounter? = with(SkikoProperties) {
+    if (!SkikoProperties.fpsEnabled) return@with null
+
+    // it is slow on Linux (100ms), so we cache it. Also refreshRate available only after window is visible
+    val refreshRate by lazy { component.graphicsConfiguration.device.displayMode.refreshRate }
 
-        if (i % count == 0) {
-            val quantile = (1 - probability) / 2.0
-            val average = (1000.0 / times.average()).roundToInt()
-            val min = (1000.0 / times.quantile(1 - quantile)).roundToInt()
-            val max = (1000.0 / times.quantile(quantile)).roundToInt()
-            val probability = (100 * probability).roundToInt()
-            println("FPS $average ($min-$max $probability%)")
+    FPSCounter(
+        periodSeconds = fpsPeriodSeconds,
+        showLongFrames = fpsLongFramesShow
+    ) {
+        when {
+            fpsLongFramesShow && fpsLongFramesMillis != null -> fpsLongFramesMillis!!
```

Comment:
Somewhat confusing logic. What is intention here?

Reply:
Agree, simplified logic a little.